### PR TITLE
Issue/1511 support pip config in projects

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
Closes https://github.com/inmanta/vscode-inmanta/issues/1511

### CODE REVIEW

**Comment by Hugo-Inmanta** [comment]
I still have a failing testcase on Jenkins (that passes on my machine)

File: server/src/inmantals/lsp_types.py [hunk 1]
```diff
@@ -36,15 +40,21 @@ class LspModel(BaseModel):
 
     class Config:
         alias_generator = partial(re.sub, r"_([a-z])", lambda match: match.group(1).upper())
-        allow_population_by_field_name = True
+        if SUPPORTS_PYDANTIC_V2:
+            populate_by_name = True
+        else:
+            allow_population_by_field_name = True
```
Comment (FloLey) [inline server/src/inmantals/lsp_types.py:46]: how does this work? I mean where is this used?

**Review by FloLey — COMMENTED** [review]
see comment. Except for that it looks good to me :)

File: server/src/inmantals/server.py [hunk 2]
```diff
@@ -165,20 +172,31 @@ def set_env(set_var: abc.Mapping[str, Optional[str]]) -> None:
...
-            {
-                "PIP_INDEX_URL": urls[0],
-                "PIP_PRE": "0" if project.install_mode == module.InstallMode.release else "1",
-                "PIP_EXTRA_INDEX_URL": " ".join(urls[1:]),
-            }
-        ):
-            logger.info("Installing modules from source: %s", mod.name)
-            project.virtualenv.install_from_source([env.LocalPackagePath(mod.path, editable=True)])
+        logger.info(f"Project {project}")
+        logger.info(f"{project.__dict__=}")
```
Comment (arnaudsjs) [inline server/src/inmantals/server.py:176]: What does this look like in the log?

File: server/src/inmantals/server.py [hunk 3]
```diff
@@ -238,21 +256,33 @@ def _get_module_name():
...
+                    content["repo"] = self.handler.repos
+
+            return content
+
         logger.debug(
-            "project.yaml created at %s, repos=%s", os.path.join(inmanta_project_dir, "project.yml"), self.handler.repos
+            "project.yml created at %s, repos=%s, pip_config=%s",
+            os.path.join(inmanta_project_dir, "project.yml"),
+            self.handler.repos,
+            self.handler.pipconfig,
```
Comment (arnaudsjs) [inline server/src/inmantals/server.py:282]: I would just print the entire `project.yml` file here.

File: server/tests/test_smoke.py [hunk 4]
```diff
@@ -26,13 +26,16 @@
...
 
+import pkg_resources
 from inmanta import env
 from inmantals import lsp_types
 from inmantals.jsonrpc import JsonRpcServer
 from inmantals.server import CORE_VERSION, InmantaLSHandler
 from packaging import version
 from pkg_resources import Requirement, parse_requirements
 
+SUPPORTS_PYDANTIC_V2: bool = version.Version(pkg_resources.get_distribution("pydantic").version) >= version.Version("2.0.0.dev")
```
Comment (arnaudsjs) [inline server/tests/test_smoke.py:37]: Why do we have this twice in the code base?

File: server/tests/test_smoke.py [hunk 5]
```diff
@@ -387,6 +409,29 @@ class MyLspType(lsp_types.LspModel):
         assert v.dict() == spec_compatible
 
 
+@pytest.mark.skipif(
+    not SUPPORTS_PYDANTIC_V2,
+    reason="Only run when pydantic v1 is supported",
```
Comment (arnaudsjs) [inline server/tests/test_smoke.py:414]:
```suggestion
    reason="Only run when pydantic v2 is supported",
```

File: server/src/inmantals/server.py (hunk 3 above)
Comment (arnaudsjs) [inline server/src/inmantals/server.py:282]: Let's add the following assertion to the constructor of this class: `self.handler.pipconfig is not None`

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: server/src/inmantals/lsp_types.py (hunk 1 above)
Comment (Hugo-Inmanta) [reply server/src/inmantals/lsp_types.py:46]: You can see this in action in these tests: [test_lsp_type_serialization_pydantic_v1](https://github.com/inmanta/vscode-inmanta/blob/b99e8494eeb3a403284503b44a1130777d3ae3ab/server/tests/test_smoke.py#L393)
and [test_lsp_type_serialization_pydantic_v2](https://github.com/inmanta/vscode-inmanta/blob/b99e8494eeb3a403284503b44a1130777d3ae3ab/server/tests/test_smoke.py#L416) 
It's used to allow populating the model with either the field name or its alias (see [pydantic doc](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.populate_by_name) for details)

File: server/src/inmantals/lsp_types.py [hunk 6]
```diff
@@ -36,15 +40,21 @@ class LspModel(BaseModel):
 
     class Config:
         alias_generator = partial(re.sub, r"_([a-z])", lambda match: match.group(1).upper())
-        allow_population_by_field_name = True
+        if SUPPORTS_PYDANTIC_V2:
+            populate_by_name = True
```
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply server/src/inmantals/lsp_types.py:44]: Right, nice catch

File: server/src/inmantals/server.py (hunk 2 above)
Comment (Hugo-Inmanta) [reply server/src/inmantals/server.py:176]: Currently looks like so:
```
2023-12-19 12:32:06,612 inmantals.server         INFO    Project <inmanta.module.Project object at 0x7f86dfcdd550>
2023-12-19 12:32:06,612 inmantals.server         INFO    project.__dict__={'_path': '/tmp/tmp0l15cjyv', '_metadata': ProjectMetadata(requires=[], name='Temporary project', description='Temporary project', freeze_recursive=False, freeze_operator='~=', author=None, author_email=None, license=None, copyright=None, modulepath=['/tmp/tmp0l15cjyv/libs'], repo=[], downloadpath='libs', install_mode=<InstallMode.release: 'release'>, relation_precedence_policy=[], strict_deps_check=True, agent_install_dependency_modules=False, pip=ProjectPipConfig(index_url='https://pypi.org/simple/', extra_index_url=[], pre=None, use_system_config=False)), 'name': 'Temporary project', 'project_path': '/tmp/tmp0l15cjyv', 'main_file': 'main.cf', '_ast_cache': None, 'module_source': <inmanta.module.ModuleV2Source object at 0x7f86dfd02cd0>, 'module_source_v1': <inmanta.module.ModuleV1Source object at 0x7f86e12bb110>, 'virtualenv': <inmanta.env.ActiveEnv object at 0x7f86e0083290>, 'loaded': False, 'modules': {}, 'root_ns': __root__, 'autostd': True, 'strict_deps_check': True, '_complete_ast': None}
```

I'll remove the first line (`logger.info(f"Project {project}")`)
And move the second one to debug level

File: package.json [hunk 7]
```diff
@@ -61,6 +61,32 @@
...
                     "scope": "resource"
                 },
+                "inmanta.pip.pre": {
+                    "type": "boolean",
+                    "default": null,
+                    "description": "Allow pre-releases when installing Python packages. Will only be used by language server if inmanta-core>=11.0.0",
+                    "scope": "resource"
+                },
+                "inmanta.pip.index_url": {
+                    "type": "string",
```
Comment (arnaudsjs) [inline package.json:71]:
```suggestion
                    "type": "string",
                    "default": null,
```

File: server/src/inmantals/server.py [hunk 8]
```diff
@@ -238,21 +257,32 @@ def _get_module_name():
...
+
+            return content
+
+        project_yml_path = os.path.join(inmanta_project_dir, "project.yml")
+        with open(project_yml_path, "w+") as fd:
+            content = _generate_project_yml()
+            yaml.dump(content, fd)
+            logger.debug(
+                f"project.yml created at {project_yml_path}:\n{content}",
+            )
```
Comment (arnaudsjs) [inline server/src/inmantals/server.py:285]:
```suggestion
            logger.debug(
                "project.yml created at %s: %s", project_yml_path, content},
            )
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

**Pull request merged**